### ISSUE TITLE
Swift 3 branch dropping iOS 8 support

### ISSUE THREAD

**Issue by tieleman** [Alamofire/Alamofire#1367]
With the coming release of Xcode 8/iOS 10/Swift 3 a lot of developers will probably move to the new Swift syntax when updating their apps, but this will most likely also include apps that still require iOS 8 compatibility. By making the Swift 3 branch iOS9+-only this will make it not easy for developers to switch over, especially since Apple is using a very aggressive iOS release cycle:

iOS 8: Sep 17, 2014
iOS 9: Sep 16, 2015
iOS 10: Sep 2016(?)

Is there a big reason to drop iOS 8 support now? I understand the need to keep the code clean and up-to-date, but I don't think iOS 8 compatibility is something that will disappear overnight for most developers (I am aware that all iOS 8 devices can be updated to iOS 9, but that's not something everyone does).

It would be very nice to have at least an initial release using Swift 3 that's compatible with iOS 8 and then moving away from iOS 8 for next releases (or make the Swift 3 branch require iOS 8.4 if it's mostly related to bugs in 8.1 and 8.2).

FWIW: the current head in the swift3 branch still compiles for iOS 8.

**Comment by tinvyhx** [comment]
The same problem .
Anyone has the idea?

**Comment by cnoon** [comment]
Thank you for all the reasoning you put together in this issue @tieleman. 

While we completely understand you point of view (believe me, we have discussed this at great length), we respectfully disagree with this approach. Apple recommends that developers always support the two latest operating systems. With a library as large and widely used as Alamofire, we need to make sure we are continually re-evaluating our tech debt and moving the needle forward while maintaining the best test suite possible. Because of this, Alamofire 4.x (Swift 3 only) will support iOS 9.0+, macOS 10.11+, tvOS 9.0+ and watchOS 2.0+. This allows us to continue to clean up our tech debt and run our test suite against all supported versions.

With that said, we're absolutely committed to still supporting iOS 8.x for some time. The Alamofire 3.x releases will fully support both Swift 2.2 AND 2.3 for the foreseeable future. The Alamofire 3.x versions will continue to receive bug fixes and security patches while the main focus of the team will be on Alamofire 4.x from here until the GM drops. When app developers are ready to embrace Swift 3, we feel that the majority of apps will be leaving behind support for iOS 8 deployment targets ([mixpanel](https://mixpanel.com/trends/#report/ios_9) currently reports iOS 9 adoption at over 92%). Now we realize that not everyone will do this, but we strongly believe that the majority of Alamofire users will. Even if you're one of the few, you do still have options.

1. You can continue to use Alamofire 3.x with Swift 2.3 until you're ready to drop iOS 8.x support. (this is our recommendation)
1. You could fork Alamofire 4.x and drop the deployment target until you're ready to move.

Alamofire is already segmented across three versions of Swift and 2-3 major versions of 4 different platforms spread over 3 different libraries. This is a tremendous amount of logic to support as well as keep the test suites up and running day-to-day while we try to iterate on Alamofire 4. With our small team, we need to draw the line eventually. There are always trade offs, and we've tried to make the best decisions for the community.

We're also working on publishing an Alamofire 4 evolution document with the proposals we'd like to see make it in over the next few weeks. Hopefully between the info provided in this response alongside the evolution doc, our team will have provided enough transparency and guidance to help the community move between Swift and Alamofire versions successfully and without difficulty.

Cheers. 🍻

**Comment by tieleman** [comment]
Thanks for the explanation, I'm sure I'll find a suitable workaround (currently I'm fixed on the commit just before the version bump), my main concern was that the move to Swift 3 doesn't mandate moving to iOS9+, so it might be an issue for other developers as well.

Also, can you link to something that shows that Apple recommends supporting only the current and previous iOS version. I did a quick Google but came out empty (might be useful to help convince clients).

Hopefully this will all be a thing of the past soon(ish):

https://lists.swift.org/pipermail/swift-dev/Week-of-Mon-20160725/002520.html

> To make this more concrete, suppose an application is written to use Swift 4, but uses packages via SwiftPM that are written using Swift 3. A single compiler would build both the app and the packages — thus ensuring that all the compiled sources are binary compatible. It would not be the case that a framework built with the Swift 3 compiler could be used by an app built using the Swift 4 compiler. That kind of library binary stability (ABI) will be a key goal of the Swift 4 release.

**Comment by cnoon** [comment]
The recommendation came from Apple in "I believe" WWDC 2014 in one of the bigger talks. I searched around [ASCIIWWDC](http://asciiwwdc.com), but I couldn't find what I was looking for. 

If I do end up tracking it down, I'll be sure to post the link here. If anyone else in the community was at the same talk I was and knows which one it was and could post the info here, it would be much appreciated.

**Comment by jamesk-au** [comment]
The same recommendation was made by Woody Lidstone in [WWDC 2016 Session 213](https://developer.apple.com/wwdc16/213), "Improving Existing Apps with Modern Best Practices":

> What are some things that you can start doing to get ready for the fall?
> 
> First of all, if you were to support a deployment target of iOS 8 and 9, you would cover 95% of the devices.
> 
> So, there's really no need any more to have any of those deployment targets set to 7.
> 
> Let's get rid of that.
> 
> In fact, the general idea is that you take the current shipping version, so 9.3, and then set your deployment target to one version back, maybe 8.4.
> 
> But don't go to 8.3 or 8.2 anymore because then you and your customers don't benefit from the improvements that we've implemented under the hood in that 8.4 release.
> 
> **When we come out with iOS X in the fall for the public, set your deployment target to 9.3.**

Also note that iOS 9 adoption was [at 86% on 18 July 2016](https://developer.apple.com/support/app-store/) (not > 90% per mixpanel).

**Comment by tsheaff** [comment]
Adding my voice to this discussion, I'd like to say that I strongly disagree with the decision to no longer support iOS 8.  In particular, you state:

```
When app developers are ready to embrace Swift 3, we feel that the majority of apps will be leaving behind support for iOS 8 deployment targets 
```

With iOS 9 support at 86% as @jamesk-au states, why would you assume major developers would be willing to make the switch so early. Dropping support for 14% of a large userbase is likely to have a huge impact on revenues.  I understand what you're saying about supporting two versions at a time, but iOS 10 isn't even out yet until later this fall. It seems reasonable to maintain iOS 8 support for a few months at minimum.

**Comment by jshier** [comment]
iOS 10 will be released one week from today. Additionally, anyone running iOS 8 can upgrade to iOS 9 for free, which they should be doing anyway, due to the rather severe security issues patched in 9.3.5. Finally, you can continue using Alamofire with the upcoming 3.5.0 release which will support iOS 8 - 10 by use of Swift 2.3.

**Comment by tsheaff** [comment]
OK cool thought iOS 10 was coming out in October. Thanks for that update.

I have one question: For Alamofire 3.x with Swift 2.3, if I'm on Cocoapods, will this play nicely with my other pods (and my main project) that do use Swift 3? If so, I can't find any good documentation on whether I need to do anything explicit in my Podfile to support different versions of swift for my different Pods.  I currently have this at the end of my Podfile to avoid having to do `Xcode > Edit > Convert > To Current Swift Syntax...` every time I run `pod install`:

```ruby
post_install do |installer|
    installer.pods_project.targets.each do |target|
        target.build_configurations.each do |config|
            config.build_settings['SWIFT_VERSION'] = '3.0'
        end
    end
end
```

I'm assuming this will need to change in some way to support some Swift 3 pods and other Swift 2.3 pods. Any guide or reference you can point us to @jshier @cnoon ? Thanks in advance.

**Comment by jshier** [comment]
No, targets must be all 2.3 or all 3, due to ABI incompatibility. I think you may be able to run a Swift 2.3 app with a Swift 3 watch app but not within the same app.

**Comment by vfn** [comment]
Just in case anyone is interested, I have forked and created a brach that adds back `iOS 8` support. I had to remove support to `URLSessionStream*` for that tho.

https://github.com/Homely/Alamofire/tree/ios8

**Comment by jshier** [comment]
@vfn If you're going to maintain a fork with iOS 8 compatibility, you should also port over the various bug fixes and workarounds that were put in for bugs in for 8.1 and 8.2, as well as proper testing support for them as well.

**Comment by vfn** [comment]
@jshier I don't have any plans to actively maintain that fork. I created it to postpone the need to drop support to iOS 8 on the app I'm working at the moment, although I'll be doing that in the near future.

I shared it here so other devs can use it as well, without having to do the same work I had already done.

I think I'll update the README file on that branch to make it clear.

**Comment by sai-prasanna** [comment]
@cnoon Why can't we prefix the classes / methods using iOS9+ API with #available() check?

I believe https://github.com/Homely/Alamofire/commit/a0608c2db96c7d98d066f8fd72c5f19b89339fc2

only the streaming classes are affected? or Do we depend on iOS9+ API for basic functionality?

**Comment by cnoon** [comment]
I'm going to lock this thread down. We've given many justifications as to why the deployment targets have been set to what they were in several different tickets. Please read through the justifications of all the tickets linked to this one if you're having issues.
